File: Arrays/arrayOfProducts.js
//! O(n) time | O(n) space
function arrayOfProducts(array) {
  const products = new Array(array.length).fill(1);

	let leftRunningProduct = 1;
	for(let i = 0; i < array.length; i++){
		products[i] = leftRunningProduct;
		leftRunningProduct *= array[i];
	}
	let rightRunningProduct = 1;
	for(let i = array.length - 1; i > -1; i--) {
		products[i] *= rightRunningProduct;
		rightRunningProduct *= array[i];
	}
}

//! O(n) time | O(n) space
function arrayOfProducts(array) {
 const products = new Array(array.length).fill(1);
 const leftProducts = new Array(array.length).fill(1);
 const rightProducts = new Array(array.length).fill(1);

	let leftRunningProduct = 1;
	for(let i = 0; i < array.length; i++) {
		leftProducts[i] = leftRunningProduct;
		leftRunningProduct *= array[i];
	}
	let rightRunningProduct = 1;
	for(let i = array.length - 1;  i > -1; i--) {
		 rightProducts[i] = rightRunningProduct;
		 rightRunningProduct *= array[i];
	}
	for(let i = 0; i < array.length; i++) {
		products[i] = leftProducts[i] * rightProducts[i];
	}
	return products;
}

//! O(n^2) time | O(n) space
function arrayOfProducts1(array) {
	const products = [];

	for(let i = 0; i <array.length; i++) {
		let runningProduct = 1;
		for(let j = 0; j < array.length; j++) {
			  if(i != j) {
					runningProduct *= array[j];
				}
		}
		products[i] = runningProduct;

	}
	return products;
}

console.log(arrayOfProducts([1, 2, 3, 4]));

File: Arrays/ArrayOfProducts.java
package CarryForward;

/**
 *

 Write a function that takes in a non-empty array of integers and returns an array of the same length, where each element in the output array is equal to the product of every other number in the input array.

 In other words, the value at output[i] is equal to the product of every number in the input array other than input[i].

 Note that you're expected to solve this problem without using division.
 Sample Input

 array = [5, 1, 4, 2]

 Sample Output

 [8, 40, 10, 20]
 // 8 is equal to 1 x 4 x 2
 // 40 is equal to 5 x 4 x 2
 // 10 is equal to 5 x 1 x 2
 // 20 is equal to 5 x 1 x 4

 */

public class ArrayOfProducts {
    public static void main(String[] args){
        int[] array = {5, 1, 4, 2};
        System.out.println(solve(array));
    }

    public static int[] solve(int[] array) {
        int[] products = new int[array.length];

        int leftRunningProduct = 1;
        for(int i = 0; i < array.length; i++) {
            products[i] = leftRunningProduct;
            leftRunningProduct *= array[i];
        }

        int rightRunningProduct = 1;
        for(int i = array.length - 1; i > -1; i--) {
            products[i] *= rightRunningProduct;
            rightRunningProduct *= array[i];
        }
        return products;
    }
}